### PR TITLE
Add homepage UI tests to addons-server

### PR DESCRIPTION
**Pull request by jrbenny35**
r? @davehunt @EnTeQuAk 

This is the first test I am bringing over from the addons-test repo. I had to change some unicode strings to work on both a local env and travis. I only brought over the functions I needed for now instead of just a cold copy + paste. 

Flake8 is failing due to the new tests but I think it would be best to ignore the directory, as formatting these to be PEP8 looks worse.

### CODE REVIEW

**Comment by jrbenny35** [comment]
@EnTeQuAk I know you don't like sudo within travis but it is needed as the trusty container image doesn't include things like MySQL. I am sure we can figure out something as I don't need to install gcc when using the trusty build.

**Comment by eviljeff** [comment]
Your flake8 error is because you commented out the code in `src/olympia/landfill/management/commands/generate_themes.py` but didn't remove the import it used.

**Comment by jrbenny35** [comment]
> Your flake8 error is because you commented out the code in src/olympia/landfill/management/commands/generate_themes.py but didn't remove the import it used.

Ah, forgot about that. I didn't even look heh, thanks for checking that out.

File: src/olympia/landfill/generators.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def create_theme(name, **extra_kwargs):
     return theme
 
 
-def generate_themes(num, owner):
+def generate_themes(num, owner, app=None):
```
Comment (eviljeff) [inline src/olympia/landfill/generators.py]: The method doesn't use `app`, so why add it as a parameter?

File: .travis.yml [hunk 2]
```diff
@@ -39,6 +39,8 @@ addons:
 before_install:
   - scripts/travis_es.sh
   - /tmp/elasticsearch/elasticsearch-1.6.2/bin/elasticsearch -d -D es.path.data=/tmp -D es.gateway.type=none -D es.index.store.type=memory -D es.discovery.zen.ping.multicast.enabled=false
+  - /sbin/start-stop-daemon --start --quiet --pidfile /tmp/custom_xvfb_99.pid --make-pidfile --background --exec /usr/bin/Xvfb -- :99 -ac -screen 0 1920x1080x16
+  - sleep 10
```
Comment (eviljeff) [inline .travis.yml]: what's this for?

**Comment by diox** [comment]
> @EnTeQuAk I know you don't like sudo within travis but it is needed as the trusty container image doesn't include things like MySQL. I am sure we can figure out something as I don't need to install gcc when using the trusty build.

The fact that the trusty image doesn't include MySQL sucks. Do you really need that image ?

File: .travis.yml [hunk 3]
```diff
@@ -1,5 +1,6 @@
+dist: trusty
 language: python
-sudo: false
+sudo: required
```
Comment (davehunt) [inline .travis.yml]: There's a [sudo-less beta](https://docs.travis-ci.com/user/trusty-ci-environment/#Using-Trusty) for trusty.. did you try this?

File: .travis.yml (hunk 2 above)
Comment (davehunt) [inline .travis.yml]: Looks like it's starting xvfb with a custom resolution. We should probably add a comment here.

File: tests/ui/conftest.py [hunk 4]
```diff
@@ -9,11 +9,21 @@
...
 from fxapom.fxapom import DEV_URL, PROD_URL, FxATestAccount
 from olympia.amo.tests import create_switch
+from selenium.webdriver.common.action_chains import ActionChains
+
 from pytest_django import live_server_helper
 
 
+@pytest.fixture
+def selenium(selenium):
+    selenium.implicitly_wait(10)
```
Comment (davehunt) [inline tests/ui/conftest.py]: We don't want to mix implicit and explicit waits. Please remove this and make sure you use explicit waits wherever they're necessary.

File: tests/ui/conftest.py [hunk 5]
```diff
@@ -9,11 +9,21 @@
...
+    selenium.set_window_size(1920, 1080)
```
Comment (davehunt) [inline tests/ui/conftest.py]: Is this a minimum required resolution, or was this just for debug purposes?

File: tests/ui/conftest.py [hunk 6]
```diff
@@ -9,11 +9,21 @@
...
+    ActionChains(selenium).reset_actions()
```
Comment (davehunt) [inline tests/ui/conftest.py]: What does this do?

File: tests/ui/conftest.py [hunk 7]
```diff
@@ -9,11 +9,21 @@
...
+    return selenium
+
+
 @pytest.fixture(scope='function')
-def my_base_url(base_url, request):
+def my_base_url(base_url, request, live_server):
```
Comment (davehunt) [inline tests/ui/conftest.py]: This means running against a remote server will attempt to start up a live server instance, which we don't want.

File: tests/ui/conftest.py [hunk 8]
```diff
@@ -49,12 +59,58 @@ def jwt_secret(base_url, variables):
...
+    from olympia.landfill.collection import generate_collection
+    from olympia.constants.applications import APPS
 
     for x in range(10):
         AddonUser.objects.create(user=user_factory(), addon=addon_factory())
+        generate_collection(addon_factory(), APPS['firefox'])
+
+
+@pytest.fixture
+def gen_addons():
```
Comment (davehunt) [inline tests/ui/conftest.py]: Please add docstrings for fixtures so it's clear what they're used for. I'm assuming this generates a collection of addons, which means the name of the fixture is misleading. Also, please avoid abbreviations in function names.

File: tests/ui/conftest.py [hunk 9]
```diff
@@ -49,12 +59,58 @@ def jwt_secret(base_url, variables):
...
+    from olympia.amo.tests import addon_factory
+    from olympia.constants.applications import APPS
+    from olympia.landfill.collection import generate_collection
+
+    for x in range(80):
```
Comment (davehunt) [inline tests/ui/conftest.py]: If `x` is unused, call it `_`. See https://hackernoon.com/understanding-the-underscore-of-python-309d1a029edc

File: tests/ui/conftest.py [hunk 10]
```diff
@@ -77,14 +133,135 @@ def create_superuser(transactional_db, my_base_url, tmpdir, variables):
...
+    print('Created custom addon for testing successfully')
+
+
+@pytest.fixture
+def force_user_login():
+    """
+        Fixture for providing the user object to seleniumlogin for logging into
+        the live_server specifically. Does not run on 'dev' or 'stage'
+        environments.
+    """
```
Comment (davehunt) [inline tests/ui/conftest.py]: First line of the docstring should be a summary. See https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

File: tests/ui/pages/desktop/base.py [hunk 11]
```diff
@@ -26,13 +35,70 @@ def login(self, email, password):
     def logout(self):
         self.header.click_logout()
 
+    @property
+    def amo_logo_title(self):
+        return self.selenium.find_element(*self._amo_logo_link_locator).get_attribute('title')
+
+    def get_url_current_page(self):
+        time.sleep(1)
+        # This is a hack until selenium actually learns patience
```
Comment (davehunt) [inline tests/ui/pages/desktop/base.py]: Can you cite the issue here? We don't want to litter our tests with arbitrary sleeps, and I'd prefer not to wrap the Selenium method for obtaining the URL in our base page object.

File: tests/ui/pages/desktop/base.py [hunk 12]
```diff
@@ -26,13 +35,70 @@ def login(self, email, password):
     def logout(self):
         self.header.click_logout()
 
+    @property
+    def amo_logo_title(self):
+        return self.selenium.find_element(*self._amo_logo_link_locator).get_attribute('title')
```
Comment (davehunt) [inline tests/ui/pages/desktop/base.py]: I don't personally think this is a good thing to check directly. Perhaps the page could provide an indirect `current_application` property that determines this based on the logo alt text, page title, or logo title as you have here.

**Comment by jrbenny35** [comment]
>I only got so far before I realised this patch is over two thousand lines, and more than 20 tests! Please keep your patches smaller so we can iterate over them much quicker during the review.

Smaller? This is just once class of tests. It has 23 tests in it however.

Could you be more specific about the best practices? I have updated all the tests to use PyPOM. I didn't see much else that was old, but maybe I don't know what you are referring too.

**Comment by davehunt** [comment]
> Smaller? This is just once class of tests. It has 23 tests in it however.

Exactly. There's also over 2,000 lines of code to review! 😨

> Could you be more specific about the best practices?

From what I saw: mixing implicit and explicit waits, sleeps, asserting the current URL or page title. I didn't review the entire pull request though. Starting smaller would make it easier to explain these individually as they come up.

**Comment by davehunt** [comment]
Oh, also we don't need to use classes for test organisation. Maybe 23 tests in a single module means there's too many tests, or we could parameterise, or we should split them into smaller modules.

**Comment by jrbenny35** [comment]
So how should I move forward?

**Comment by jrbenny35** [comment]
I honestly did not think I changed 2000 lines. I can use pypoms ```wait_for_page_to_load``` instead of checking the url. Some tests do require the url check though. I

**Comment by jrbenny35** [comment]
Closing in favor of a major overhaul of these tests.

**Pull request closed without merging**